File: 프로그래머스/0/181918. 배열 만들기 4/배열 만들기 4.py
def solution(arr):
    stk = []
    
    idx = 0
    stk.append(arr[idx])
    
    idx = 1
    
    # for idx in range(1, len(arr)):
    while idx < len(arr):
        if len(stk) == 0:
            stk.append(arr[idx])
            idx += 1
            
        elif (stk[-1] < arr[idx]):
            stk.append(arr[idx])
            idx += 1
        
        else:
            stk.pop()
            
    return stk

File: 프로그래머스/0/181918. 배열 만들기 4/배열 만들기 4.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> arr) {
    vector<int> stk;
    
    int idx = 0;
    stk.push_back(arr[idx]);
    
    idx = 1;
    
    while (idx < arr.size()) {
        int len = stk.size();
        if (len == 0) {
            stk.push_back(arr[idx]);
            idx += 1;
        }
        else if (stk[len-1] < arr[idx]) {
            stk.push_back(arr[idx]);
            idx += 1;
        }
        
        else
            stk.pop_back();
    }
    return stk;
}